Session Controller: data logging controller class

cosparks opened the pull request:
Most of the info explaining this class is in #73:

```SessionController``` class is the **abstract** base class for 'session' (data logging) modes of dyno.  This class will implement all functionality associated with logging data and will be configurable via the ```StateInfo``` object passed to its ```init``` method, which contains:
- A logging interval: this dictates how often we will be logging values to the output csv file
- An output filename: the name of the csv file which ```SessionController``` will log values to when it is running
- A list of ```ValueId```: ```ValueId```s are enum values, each of which uniquely represents one of the loggable properties of the dyno (angular velocity, rpm, force, angular acceleration, etc..), and will be used to configure the list of logging functions, the column names for the csv header and also the values which will be displayed on the screen

Currently, this class is missing some of the logging functions, all of which have been marked with TODOs.  I think that implementing these here is beyond the scope of this PR.

closes #73

cosparks committed:
Adds SessionController class to project

cosparks committed:
Merge branch 'dev-valueids' into dev-session

cosparks committed:
Adds SessionController to project

cosparks committed:
Merges changes from dev -- removes comments from application.h

cosparks committed:
Merge branch 'dev' into dev-session

cosparks committed:
Fixes build issue

cosparks committed:
removes unnecessary initialization of value in SessionController

cosparks committed:
Removes input handlers from abstract SessionController class

LaPommeCosmique commented on src/application/SessionController.cpp:
```diff
@@ -0,0 +1,160 @@
+    _info = info;
+}
+
+void SessionController::_logValues() {
```
what do you think about adding an assert here to verify that the columnNum of dataLogger is 0? Just to make sure that we are still lined up properly

cosparks replied:
I had the same thought, but if this does somehow go out of sync then it would spam the error log file because _logValues will be called many thousands of times while logging.

There are already tests for CsvFile::addEntry, so I trust that it won't fall out of line.  I have done some manual testing as well.  What I could do is make _outputCsv a private field so that it can only be controlled via protected SessionController methods.  It might also be worthwhile adding a unit test to make sure that everything is initialized properly.

LaPommeCosmique replied:
> What I could do is make _outputCsv a private field

I would be on board with this. But if we are going this direction, if we want to be completely safe I believe we would also have to restrict access to 
```
protected:
std::vector<std::function<String(void)>> _valueLoggers; // tandem with _valueIds
std::vector<ValueId> _valueIds;                         // tandem with _valueLoggers
```
to prevent messing with the sizes of the valueLoggers vector

LaPommeCosmique replied:
Okay, now that I think about it I think just making _outputCsv private is enough lol. I dont think it will be likely that people will be changing _valueLoggers after initialization.

cosparks replied:
I still haven't built out the derived classes, but it might be worth making some changes to SessionController once those take shape.  In the meantime, _valueIds and _valueLoggers may be needed to update display values in derived classes.  I'll definitely revisit this after building out the calibration controller class

cosparks committed:
Set outputCsv to private and remove (1) init, (2) _countEntries and (3) default logging interval
- (1) makes clearer the intention that SessionController isn't an instantiable class
- (2) SessionController::_countEntries was no longer being used
- (3) now requires that there always be a valid value in state info object for

cosparks commented:
Ok, I've made ```SessionController::_outputCsv``` a private field, and I also moved ALL initialization behaviour out of ```SessionController::init``` and into ```SessionController::_initializeOutput```, which I think makes clearer exactly what SessionController is supposed to be doing.
